### PR TITLE
IB: Support ConnectX-3 MLNX cards

### PR DESCRIPTION
**Pull request by kamalca**
Older cards such as ConnectX-3 require legacy LTS OFED drivers.

### CODE REVIEW

**Comment by kamalca** [comment]
This change will still have errors for certain hardware / OS combinations. The LTS drivers for ConnectX-3 cards do not support newer OS distros.

File: lisa/features/infiniband.py [hunk 1]
```diff
@@ -146,6 +149,8 @@ def _get_ib_device_names(self) -> List[str]:
     def _get_mofed_version(self) -> str:
         node = self._node
         default = "5.4-3.0.3.0"
+        if self._is_legacy_device():
+            return "4.9-5.1.0.0"
```
Comment (squirrelsc) [inline lisa/features/infiniband.py:155]: Is it possible to parse out the version directly?

File: lisa/sut_orchestrator/azure/features.py [hunk 2]
```diff
@@ -563,6 +563,11 @@ def is_over_nd(self) -> bool:
         dmesg = self._node.tools[Dmesg]
         return "hvnd_try_bind_nic" in dmesg.get_output()
 
+    def _is_legacy_device(self) -> bool:
+        lspci = self._node.tools[Lspci]
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/features.py]: Why is this method implemented in Azure? Shouldn't be it independent with cloud?

**Comment by squirrelsc** [comment]
> This change will still have errors for certain hardware / OS combinations. The LTS drivers for ConnectX-3 cards do not support newer OS distros.

What can we do on this?

File: lisa/sut_orchestrator/azure/features.py (hunk 2 above)
Comment (kamalca) [reply lisa/sut_orchestrator/azure/features.py]: I wasn't sure whether this is cloud independent. Other clouds may use other legacy cards such as ConnectX-IB which I don't think are used in Azure. But I guess the same logic could still be used across all clouds. I can move this logic to infiniband.py.

File: lisa/features/infiniband.py (hunk 1 above)
Comment (kamalca) [reply lisa/features/infiniband.py:155]: This is not parsing the version on the VM, this is making a decision for which OFED drivers to install.

Most stable is 5.4 (LTS), old cards require 4.9 LTS, newest distros may require newer versions

**Comment by kamalca** [comment]
> > This change will still have errors for certain hardware / OS combinations. The LTS drivers for ConnectX-3 cards do not support newer OS distros.
> 
> What can we do on this?

We can't control which distros NVIDIA supports with the 4.9 LTS drivers. But I could add some logic to raise an UnsupportedDistro / Skipped exception for these cases.

File: lisa/features/infiniband.py [hunk 3]
```diff
@@ -305,15 +314,19 @@ def setup_rdma(self) -> None:
...
-                f"--skip-repo --skip-unsupported-devices-check --without-fw-update"
+                f"--skip-repo --without-fw-update"
             )
 
+        if not self._is_legacy_device():
+            extra_params += " --skip-unsupported-devices-check"
+
         node.execute(
-            f"./{mofed_folder}/mlnxofedinstall --add-kernel-support {extra_params}",
+            f"./{mofed_folder}/mlnxofedinstall --add-kernel-support " f"{extra_params}",
```
Comment (squirrelsc) [inline lisa/features/infiniband.py]: This line doesn't need to be changed.

**Review by squirrelsc — APPROVED** [review]

**Pull request merged**